Pull Request for "gantt" format

RacoonDev opened the pull request:
This PR is made in reference to: #364

This PR addresses or contains:
- Provides a "gantt" result printer making use of the Mermaid library

This PR includes:
- [ ] Update of RELEASE-NOTES.md
- [x] Tests (unit/integration)
- [x] CI build passed

Fixes #

kghbln commented:
Thanks a lot for the pull. I have requested a review.

kghbln commented on DefaultSettings.php:
```diff
@@ -53,7 +53,7 @@
 	'datatables',
 	'spreadsheet',
-
+	'gantt'
```
A blank line between this and the following note as before will be fine

RacoonDev replied:
Because of consitency removed the blank line
See: 56

Commit: 3aeae38c7300a22b8e6697a7fac90ef400a5a0b4

kghbln commented on i18n/de.json:
```diff
@@ -337,5 +337,19 @@
 	"srf-paramdesc-spreadsheet-filename": "Das zu erzeugende Dateiformat. Erlaubte Werte: xlsx, xls, ods, csv. Standard: xlsx",
 	"srf-paramdesc-spreadsheet-templatefile": "Der Name einer Tabellenkalkulationsdatei aus dem Namensraum ''Datei'', die zur Formatierung der erzeugten Datei verwendet wird.",
-	"srf-paramdesc-icalendar-timezone": "Legt eine durch Kommata getrennte Liste mit Zeitzonen fest"
+	"srf-paramdesc-icalendar-timezone": "Legt eine durch Kommata getrennte Liste mit Zeitzonen fest",
```
Usually there is no need to provide I18n other than en and qqq, however since it is there it's ok.

kghbln commented on i18n/en.json:
```diff
@@ -337,5 +337,19 @@
+	"srf-error-charttype": "The '''charttype''' you have choosen is not supported",
+	"srf-error-theme" : "The '''theme''' you have choosen is not supported",
+	"srf-error-sortkey" : "The '''sortkey''' you have choosen is not supported",
+	"srf-printername-mermaid": "Mermaid Extension"
```
This holds the printername shown e.g. for selection special page "Ask". Should just be "Gantt" or probably better "Gantt chart" instead of "Mermaid Extension".

RacoonDev replied:
changed printername to "Gantt Chart"

kghbln commented on i18n/qqq.json:
```diff
@@ -353,5 +353,14 @@
+	"srf-paramdesc-titletopmargin": "{{doc-paramdesc|titletopmargin}}",
+	"srf-paramdesc-barheight": "{{doc-paramdesc|barheight}}",
+	"srf-paramdesc-leftpadding": "{{doc-paramdesc|leftpadding}}",
+	"srf-paramdesc-bargap": "{{doc-paramdesc|bargap}}"
```
Some docus are missing, e.g.
```
"srf-error-mapping-assignment": "This is an error message.\n\nParameter:\n* $1 holds ..."

"srf-error-mapping-keywords": "This is an error message."
"srf-printername-mermaid": "{{doc-smwformat|Gantt}}"
```
Sometimes the docu can be rather simple but still providing some context.

RacoonDev replied:
Added missing docus on Commit:a7c6754d9d24145679a0336343024b86d2da82ca

kghbln commented on i18n/qqq.json:
```diff
@@ -353,5 +353,19 @@
+	"srf-paramdesc-gantt-bargap": "{{doc-paramdesc|bargap}}",
+	"srf-error-gantt-theme" : "This is an error message.",
+	"srf-error-gantt-sortkey" : "This is an error message.",
+	"srf-error-gantt-mapping-assignment": "This is an error message.",
```
The docu for $1 still missing per my previous suggestion.

kghbln commented on i18n/qqq.json:
```diff
@@ -353,5 +353,19 @@
+	"srf-error-gantt-sortkey" : "This is an error message.",
+	"srf-error-gantt-mapping-assignment": "This is an error message.",
+	"srf-error-gantt-mapping-keywords": "This is an error message.",
+	"srf-printername-gantt": "{{doc-smwformat|vcard}}"
```
I guess this is meant to be gantt

RacoonDev replied:
Commit-ID: 8f01390aeb3ff92489fa8758f44a1b7d593d8f8a

s7eph4n commented:
I have one more question: Why do you need the `sortkey` parameter? SMW already has `sort` and as far as I can tell it should work fine.

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+	protected $mErrors = [];
+
+	public function getName() {
+		// Give grep a chance to find the usage
```
Commit-ID: 675be2d0564d19a6ea9efda50383528fe1e6d03c

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+		$params[] = [
+			'name'    => 'titletopmargin',
+			'message' => 'srf-paramdesc-titletopmargin',
+			'default' => 25
```
Fixed in lines 40-91

Commit-ID: 675be2d0564d19a6ea9efda50383528fe1e6d03c

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+		$this->mParams['prioritymapping'] = trim( $params['prioritymapping'] );
+		$this->mParams['theme'] = trim( $params['theme'] );
+
+		//Validate Theme
```
Out source the validation to own function
See line 116 and 282-297

Commit: 677eab9a94583bae18fa8363301ccd3412105785

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+					// if no key value pair
+					if ( count( $pmKeyVal ) % 2 != 0 ) {
+						array_push( $this->mErrors,
+							wfMessage( 'srf-error-mapping-assignment', 'statusmapping' )->text() );
```
I rewrite the hole mechanism that takes care of the mapping validation. Within this change i changed the message key.
Commit: 675be2d0564d19a6ea9efda50383528fe1e6d03c

RacoonDev commented on formats/Gantt/src/GanttTask.php:
```diff
@@ -0,0 +1,110 @@
+			$mapping = [];
+
+			foreach ( $paramMapping as $pm ) {
+				$pmKeyVal = explode( '=>', $pm );
```
Line 133:
Commit: 675be2d0564d19a6ea9efda50383528fe1e6d03c

RacoonDev commented on formats/Gantt/src/GanttTask.php:
```diff
@@ -0,0 +1,110 @@
+					if ( $type == "status" ) {
+						$this->setStatus($realParam);
+					} else {
+						if ( $type == "priority" ) {
```
Rewrite setTaskParam fn 
See line 63-83

Commit: 66c43435712459a6509ea9677c833cb00f7cbdd9

RacoonDev commented on tests/phpunit/Unit/Formats/MermaidTest.php:
```diff
@@ -0,0 +1,30 @@
+	/**
+	 * @see QueryPrinterRegistryTestCase::getFormats
+	 *
+	 * @since 1.8
```
Should this be the version of SMW on when the new format was implemented?

RacoonDev commented on tests/phpunit/Unit/Mermaid/GanttTaskTest.php:
```diff
@@ -0,0 +1,32 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Test of core classes were moved to integration tests

RacoonDev commented on tests/phpunit/Unit/Mermaid/GanttSectionTest.php:
```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Test of core classes were moved to integration tests

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+				if ( !strpos( $pm, '=>' ) ) {
+					array_push( $this->mErrors, wfMessage( 'srf-error-mapping-assignment', 'statusmapping' )->text() );
+				} else {
+					$pmKeyVal = explode( '=>', $pm );
```
See lines 129-162

Commit: 675be2d0564d19a6ea9efda50383528fe1e6d03c

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+						array_push( $this->mErrors,
+							wfMessage( 'srf-error-mapping-assignment', 'statusmapping' )->text() );
+					} else {
+						$mapping[trim( $pmKeyVal[0] )] = trim( $pmKeyVal[1] );
```
Line 126

Commit:675be2d0564d19a6ea9efda50383528fe1e6d03c

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+		if ( !class_exists( 'Mermaid' ) && !class_exists( 'Mermaid\\MermaidParserFunction' ) ) {
+			//wfWarn( 'The SRF Mermaid format needs the Mermaid extension to be installed.' );
+		//catch ( Exception $exception ) {
+				$queryResult->addErrors( ["Error: Mermaid Extension needs to be installed."] );
```
Line 182

Commit: a32e25f13f125c753a9891bb9cdcd22935a0ba47

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+			}
+		}
+
+		$this->mGantt = new Gantt( $this->mParams );
```
Line 168 and 293-296
Commit: a32e25f13f125c753a9891bb9cdcd22935a0ba47

RacoonDev commented on Resources.php:
```diff
@@ -1012,6 +1012,11 @@
 
+	// Mermaid Format
+	'ext.srf.gantt' => $formatModule + [
+			'scripts' => 'Gantt/resources/ext.gantt.js',
```
Line 1018

Commit: 4846b1c06df9e80397c1446110bfc49d5c053458

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+
+		// Show warning if Extension:Mermaid is not available
+		if ( !class_exists( 'Mermaid' ) && !class_exists( 'Mermaid\\MermaidParserFunction' ) ) {
+			//wfWarn( 'The SRF Mermaid format needs the Mermaid extension to be installed.' );
```
Lines 182-183

Commit: fec7095cd06dab3cb0ab9240f36f2c0bb2c9f803

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+		SMWOutputs::requireResource( 'ext.mermaid' );
+		SMWOutputs::requireResource( 'ext.mermaid.styles' );
+		SMWOutputs::requireResource( 'ext.srf.gantt' );
+
```
Load necessary modlues via dependency.
See lines 189-190

Commit: fec7095cd06dab3cb0ab9240f36f2c0bb2c9f803

See lines 1018
Commit: 4846b1c06df9e80397c1446110bfc49d5c053458

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+		$this->mParams['sortkey'] = trim( $params['sortkey'] );
+		$this->mParams['statusmapping'] = trim( $params['statusmapping'] );
+		$this->mParams['prioritymapping'] = trim( $params['prioritymapping'] );
+		$this->mParams['theme'] = trim( $params['theme'] );
```
Lines 112-119
This also will be refactored in part of the conversation above. (Remove handleParameters and use instead getParameters)

Commit: 75e696523157f04b547068adbc8eee97d7666e9c

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+ * @ingroup SMW
+ *
+ * @licence GNU GPL v2+
+ * @author mwjames
```
Line 6:

Commit: 0afa980093650bc09e123e8ccbb42869e4ea7834

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+                var data = $(this).data('mermaid');
+
+                $(this).find('.mermaid-dots').hide();
+                $(this).append('<div id=' + id + '> ' + data.content + ' </div>');
```
If this is added on server side i will lose the line breaks "/n" so i append it via js

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+                mermaid.init(undefined, $("#" + id));
+
+                // replace <esc>35</esc> with # Tag
+                $("svg text:contains('<esc>35</esc>')").each(function () {
```
Line 28:
Commit: 0afa980093650bc09e123e8ccbb42869e4ea7834

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+        });
+    });
+
+    function forceTextWrappingOn(node, width) {
```
Lines 42-46

Commit: f58ad407f2218e501027ad349d0502cdb0005f9f

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+                var sectionTitleNodes = document.querySelectorAll('text.sectionTitle');
+                for (var n in sectionTitleNodes)
+                    if (sectionTitleNodes.hasOwnProperty(n))
+                        forceTextWrappingOn(sectionTitleNodes[n], data.config.gantt.leftPadding - 37);
```
Rewrite function
See line 34-36
Commit: 4297b4592035291c79804ce0d049749ff353dce9

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+
+                // word wrapping in <text>
+                var sectionTitleNodes = document.querySelectorAll('text.sectionTitle');
+                for (var n in sectionTitleNodes)
```
Rewrite function
See line 34-36
Commit: [4297b45](https://github.com/SemanticMediaWiki/SemanticResultFormats/commit/4297b4592035291c79804ce0d049749ff353dce9)

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+                });
+
+                // word wrapping in <text>
+                var sectionTitleNodes = document.querySelectorAll('text.sectionTitle');
```
Rewrite function
See line 34-36
Commit: [4297b45](https://github.com/SemanticMediaWiki/SemanticResultFormats/commit/4297b4592035291c79804ce0d049749ff353dce9)

RacoonDev commented on formats/Gantt/resources/ext.gantt.js:
```diff
@@ -0,0 +1,79 @@
+                    tspanWidth = tspan == null ? 0 : tspan.getComputedTextLength();
+                    if (tspanWidth > width || tspanWidth === 0) {
+                        if(index !== 0){
+                            y = y + 10;
```
Line 53 & 64

Commit: e4b2bf6d8d89069d4b2994e142c6fffdb5124c0d

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+	public function addSection( $sectionID, $sectionTitle, $startDate, $endDate, $taskID ) {
+
+		$sections = $this->getSections();
+		if ( count( $sections ) != 0 ) {
```
rewrite function.
Lines 110-127

Commit: 90005797b81fa936702b75591cf762249401d61a

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+
+			if ( array_key_exists( $sectionID, $sections ) ) {
+
+				foreach ( $sections as $sectionObj ) {
```
rewrite function.
Lines 110-127

Commit: [9000579](https://github.com/SemanticMediaWiki/SemanticResultFormats/commit/90005797b81fa936702b75591cf762249401d61a)

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+			return strcmp( $a->getTitle(), $b->getTitle() );
+		} else {
+			if ( strpos( $sortKey, 'date' ) !== false ) {
+				if ( strpos( $sortKey, 'start' ) !== false ) {
```
Remove sort functionality, this can be done with inline query standard sort and order property
Line: 126-182

Commit: 6aa6b90db221170166efcb53f983a06dfe394b03

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+		if ( !empty( $sortKey ) ) {
+			// Order Sections
+			usort( $sections, [ $this, "sortSections" ] );
+			usort( $tasks, [ $this, "sortTasks" ] );
```
Remove sort functionality, this can be done with inline query standard sort and order property
Line: 126-182

Commit: 6aa6b90db221170166efcb53f983a06dfe394b03

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+			// reorder TaskArray of current section
+			foreach ( $sections as $section ) {
+				$orderedTasks = [];
+				foreach ( $tasks as $task ) {
```
Remove sort functionality, this can be done with inline query standard sort and order property
Line: 126-182

Commit: 6aa6b90db221170166efcb53f983a06dfe394b03

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+		}
+
+		/*
+		 * Bring the "section" with no title to the first position.
```
Remove sort functionality, this can be done with inline query standard sort and order property
Line: 126-182

Commit: 6aa6b90db221170166efcb53f983a06dfe394b03

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+
+			//loop through related section tasks
+			foreach ( $section->getTasks() as $sectionTaskValue ) {
+				foreach ( $tasks as $taskObj ) {
```
Remove sort functionality, this can be done with inline query standard sort and order property
Line: 126-182

Commit: 6aa6b90db221170166efcb53f983a06dfe394b03

RacoonDev commented on formats/Gantt/src/GanttTask.php:
```diff
@@ -0,0 +1,110 @@
+			foreach ( $paramMapping as $pm ) {
+				$pmKeyVal = explode( '=>', $pm );
+				$mapping[trim($pmKeyVal[0])] = trim($pmKeyVal[1]);
+			}
```
Refactor "setTaskParams"
Line 64-73
Commit: 66c43435712459a6509ea9677c833cb00f7cbdd9

RacoonDev commented on formats/Gantt/src/Gantt.php:
```diff
@@ -0,0 +1,311 @@
+		$tasks = $this->getTasks();
+		$sortKey = $this->getSortKey();
+
+		if ( !empty( $sortKey ) ) {
```
Remove sort functionality, this can be done with inline query standard sort and order property

Commit: [6aa6b90](https://github.com/SemanticMediaWiki/SemanticResultFormats/commit/6aa6b90db221170166efcb53f983a06dfe394b03)

RacoonDev commented:
The CI tests are still failing. #The step to getting it to work is probably to include the Mermaid extension in the test setup.

RacoonDev commented:
The CI tests are still failing this is because the Mermaid Extension that is needed to succeed the tests is missing. Can pls anyone tell me how i can set a dependency to this extension in my CI tests?

RacoonDev commented on formats/Gantt/GanttPrinter.php:
```diff
@@ -0,0 +1,309 @@
+			array_push( $this->mErrors, wfMessage( 'srf-error-sortkey' )->text() );
+		}
+
+		//Validate mapping
```
Solved in Commit: e9fbc30bd12ea6b12ae43989ecd9756d98b33bf5

kghbln commented:
> The CI tests are still failing this is because the Mermaid Extension that is needed to succeed the tests is missing. Can pls anyone tell me how i can set a dependency to this extension in my CI tests?

I guess it should be something like adapting "install-semantic-result-formats.sh" to install mermaid 2.0

mwjames commented:
> I guess it should be something like adapting "install-semantic-result-formats.sh" to install mermaid 2.0

Exactly, just look around how additional dependencies can be loaded [0]. Also see [1, 2, 3] how integration tests can define a `require` for an external dependency and if it cannot the dependency then the test is skipped.

[0] https://github.com/SemanticMediaWiki/SemanticApprovedRevs/blob/master/tests/travis/install-semantic-approved-revs.sh
[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/93d4940185c30e7d7a58f6e0f7e018113fe7807d/tests/phpunit/Integration/JSONScript/TestCases/q-1300.json#L3-L5
[2] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/JSONScript/README.md#define-a-dependency
[3] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/93d4940185c30e7d7a58f6e0f7e018113fe7807d/tests/phpunit/Integration/JSONScript/JsonTestCaseScriptRunnerTest.php#L176-L196

PS: I'm fairly busy with smw-core, so I won't find time within the ext couple of weeks to have a look at it but please don't drag this too long otherwise it will code rot.

RacoonDev commented:
@mwjames, @kghbln, @s7eph4n:
Thanks for your hints. I try to fix the Travis CI build but this will trigger some pushes to the repo. Jfyi

RacoonDev commented:
@mwjames: At the moment mermaid extension runs with manifest_version: 2.  This just works with MW 1.29+. Tests execute with the following error msg: 

**Fatal error: Uncaught exception 'Exception' with message '/home/travis/build/RacoonDev/mw/extensions/Mermaid/extension.json: unsupported manifest_version: 2' in** 

to have the ability to test SRF with Mermaid Extension in MW < 1.29 we need to set the manifest_version fo Extension:Mermaid to "1".

mwjames commented:
> to have the ability to test SRF with Mermaid Extension in MW < 1.29 we need
> to set the manifest_version fo Extension:Mermaid to "1".

SRF 3.1 will switch to MW 1.31 as minimum [0], so that shouldn't pose
any issue. The travis-CI matrix should be amended outside of the PR to
meet the 3.1 requirement.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/docs/COMPATIBILITY.md#platform-compatibility-and-release-status

> cc40a937de9156cb9316be574a26bc17ffee6236  add
> build/travis/install-mermaid.sh, extend JsonTestCaseScriptRunnerTest.php

A quick look reveals:

Not sure why you copied the `runTestCaseFile` method but that doesn't
belong here.

```
	/**
	 * @see JsonTestCaseScriptRunner::runTestCaseFile
	 *
	 * @param JsonTestCaseFileHandler $jsonTestCaseFileHandler
	 */
	protected function runTestCaseFile( JsonTestCaseFileHandler
$jsonTestCaseFileHandler ) {
		$this->checkEnvironmentToSkipCurrentTest( $jsonTestCaseFileHandler );
		// Setup
		$this->prepareTest( $jsonTestCaseFileHandler );
		// Before test execution
		$this->doRunBeforeTest( $jsonTestCaseFileHandler );
		// Run test cases
		$this->doRunParserTests( $jsonTestCaseFileHandler );
		$this->doRunSpecialTests( $jsonTestCaseFileHandler );
		$this->doRunRdfTests( $jsonTestCaseFileHandler );
		$this->doRunQueryTests( $jsonTestCaseFileHandler );
		$this->doRunParserHtmlTests( $jsonTestCaseFileHandler );
		$this->doRunApiTests( $jsonTestCaseFileHandler );
	}
```

When you define a dependency, please make sure it does test against
something meaningful such as `MERMAID_VERSION` and not `SM_VERSION`.

This should have failed locally for obvious reasons!

```
	/**
	 * @see JsonTestCaseScriptRunner::getDependencyDefinitions
	 */
	protected function getDependencyDefinitions() {
		return [
			'Mermaid' => function( $val, &$reason ) {
				if ( !defined( 'SM_VERSION' ) ) {
					$reason = "Dependency: Mermaid as requirement is not available!";
					return false;
				}
				list( $compare, $requiredVersion ) = explode( ' ', $val );
				$version = SM_VERSION;
				if ( !version_compare( $version, $requiredVersion, $compare ) ) {
					$reason = "Dependency: Required version of Mermaid
($requiredVersion $compare $version) is not available!";
					return false;
				}
				return true;
			}
		];
	}
```

For better readability, you may write:

```
	/**
	 * @see JsonTestCaseScriptRunner::getDependencyDefinitions
	 */
	protected function getDependencyDefinitions() {
		return [
			'Mermaid' => [ $this, 'checkMermaidDependency' ]
		];
	}

	/**
	 * Explain the reason ...
	 */
	public function checkMermaidDependency( $val, &$reason ) {

		if ( !defined( 'MERMAID_VERSION' ) ) {
			$reason = "Dependency: Mermaid as requirement is not available!";
			return false;
		}

		list( $compare, $requiredVersion ) = explode( ' ', $val );
		$version = MERMAID_VERSION;

		if ( !version_compare( $version, $requiredVersion, $compare ) ) {
			$reason = "Dependency: Required version of Mermaid
($requiredVersion $compare $version) is not available!";
			return false;
		}

		return true;
	}
```

Cheers

On 4/1/19, Sebastian Schmid <notifications@github.com> wrote:
> @mwjames: At the moment mermaid extension runs with manifest_version: 2.
> This just works with MW 1.29+. Tests execute with the following error msg:
>
> **Fatal error: Uncaught exception 'Exception' with message
> '/home/travis/build/RacoonDev/mw/extensions/Mermaid/extension.json:
> unsupported manifest_version: 2' in**
>
> to have the ability to test SRF with Mermaid Extension in MW < 1.29 we need
> to set the manifest_version fo Extension:Mermaid to "1".
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/SemanticResultFormats/pull/467#issuecomment-478495422

mwjames commented:
> 6cb76de6c7d9ba9e1ac279d533f2faf14b590b85  remove runTestCaseFile(), rewrite
> checkMermaidDependency

Why did you remove `getPermittedSettings`? Stephan added this for a reason [0].

Also, when overriding a method from a parent class (as in case of
`getDependencyDefinitions`) make sure to check those returning values
before. For example, `parent::getDependencyDefinitions()` returns a [
... ] and you should merge them with added definitions.

Why did you add `protected function runTestCaseFile(
JsonTestCaseFileHandler $jsonTestCaseFileHandler ) { }`?

`JsonTestCaseScriptRunnerTest` is inherited from [1], so it will call
methods as needed except when they overridden. Your change actually
makes all integration tests fail because `runTestCaseFile` isn't
execute on any test because of the empty `{}`.

Did you run those tests locally? You should have noticed some issues
with those changes, I strongly suggest to run tests before posting a
PR.

[0] https://github.com/SemanticMediaWiki/SemanticResultFormats/commit/eea7866c917a766af8836c1b387ea7ced080985c
[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/JSONScript/JsonTestCaseScriptRunnerTest.php

Cheers

On 4/1/19, James HK <jamesin.hongkong.1@gmail.com> wrote:
>> to have the ability to test SRF with Mermaid Extension in MW < 1.29 we
>> need
>> to set the manifest_version fo Extension:Mermaid to "1".
>
> SRF 3.1 will switch to MW 1.31 as minimum [0], so that shouldn't pose
> any issue. The travis-CI matrix should be amended outside of the PR to
> meet the 3.1 requirement.
>
> [0]
> https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/docs/COMPATIBILITY.md#platform-compatibility-and-release-status
>
>> cc40a937de9156cb9316be574a26bc17ffee6236  add
>> build/travis/install-mermaid.sh, extend JsonTestCaseScriptRunnerTest.php
>
> A quick look reveals:
>
> Not sure why you copied the `runTestCaseFile` method but that doesn't
> belong here.
>
> ```
> 	/**
> 	 * @see JsonTestCaseScriptRunner::runTestCaseFile
> 	 *
> 	 * @param JsonTestCaseFileHandler $jsonTestCaseFileHandler
> 	 */
> 	protected function runTestCaseFile( JsonTestCaseFileHandler
> $jsonTestCaseFileHandler ) {
> 		$this->checkEnvironmentToSkipCurrentTest( $jsonTestCaseFileHandler );
> 		// Setup
> 		$this->prepareTest( $jsonTestCaseFileHandler );
> 		// Before test execution
> 		$this->doRunBeforeTest( $jsonTestCaseFileHandler );
> 		// Run test cases
> 		$this->doRunParserTests( $jsonTestCaseFileHandler );
> 		$this->doRunSpecialTests( $jsonTestCaseFileHandler );
> 		$this->doRunRdfTests( $jsonTestCaseFileHandler );
> 		$this->doRunQueryTests( $jsonTestCaseFileHandler );
> 		$this->doRunParserHtmlTests( $jsonTestCaseFileHandler );
> 		$this->doRunApiTests( $jsonTestCaseFileHandler );
> 	}
> ```
>
> When you define a dependency, please make sure it does test against
> something meaningful such as `MERMAID_VERSION` and not `SM_VERSION`.
>
> This should have failed locally for obvious reasons!
>
> ```
> 	/**
> 	 * @see JsonTestCaseScriptRunner::getDependencyDefinitions
> 	 */
> 	protected function getDependencyDefinitions() {
> 		return [
> 			'Mermaid' => function( $val, &$reason ) {
> 				if ( !defined( 'SM_VERSION' ) ) {
> 					$reason = "Dependency: Mermaid as requirement is not available!";
> 					return false;
> 				}
> 				list( $compare, $requiredVersion ) = explode( ' ', $val );
> 				$version = SM_VERSION;
> 				if ( !version_compare( $version, $requiredVersion, $compare ) ) {
> 					$reason = "Dependency: Required version of Mermaid
> ($requiredVersion $compare $version) is not available!";
> 					return false;
> 				}
> 				return true;
> 			}
> 		];
> 	}
> ```
>
> For better readability, you may write:
>
> ```
> 	/**
> 	 * @see JsonTestCaseScriptRunner::getDependencyDefinitions
> 	 */
> 	protected function getDependencyDefinitions() {
> 		return [
> 			'Mermaid' => [ $this, 'checkMermaidDependency' ]
> 		];
> 	}
>
> 	/**
> 	 * Explain the reason ...
> 	 */
> 	public function checkMermaidDependency( $val, &$reason ) {
>
> 		if ( !defined( 'MERMAID_VERSION' ) ) {
> 			$reason = "Dependency: Mermaid as requirement is not available!";
> 			return false;
> 		}
>
> 		list( $compare, $requiredVersion ) = explode( ' ', $val );
> 		$version = MERMAID_VERSION;
>
> 		if ( !version_compare( $version, $requiredVersion, $compare ) ) {
> 			$reason = "Dependency: Required version of Mermaid
> ($requiredVersion $compare $version) is not available!";
> 			return false;
> 		}
>
> 		return true;
> 	}
> ```
>
> Cheers
>
> On 4/1/19, Sebastian Schmid <notifications@github.com> wrote:
>> @mwjames: At the moment mermaid extension runs with manifest_version: 2.
>> This just works with MW 1.29+. Tests execute with the following error
>> msg:
>>
>> **Fatal error: Uncaught exception 'Exception' with message
>> '/home/travis/build/RacoonDev/mw/extensions/Mermaid/extension.json:
>> unsupported manifest_version: 2' in**
>>
>> to have the ability to test SRF with Mermaid Extension in MW < 1.29 we
>> need
>> to set the manifest_version fo Extension:Mermaid to "1".
>>
>> --
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/SemanticMediaWiki/SemanticResultFormats/pull/467#issuecomment-478495422
>

mwjames commented:
Please have look at the Travis-CI output, something you did in
`install-mermaid.sh` now makes all Travis-CI jobs to fail.

I'm not sure why you added `install-mermaid.sh` and not just amended
the existing script but with those changes we are unable to make any
successful test run.

```
Writing lock file
Generating autoload files
+cd extensions
+cd Mermaid
+'[' 467 '!=' false ']'
+git fetch origin +refs/pull/467/merge:
fatal: Couldn't find remote ref refs/pull/467/merge
The command "bash ./build/travis/install-mermaid.sh" failed and exited
with 128 during .
```

On 4/2/19, James HK <jamesin.hongkong.1@gmail.com> wrote:
>> 6cb76de6c7d9ba9e1ac279d533f2faf14b590b85  remove runTestCaseFile(),
>> rewrite
>> checkMermaidDependency
>
> Why did you remove `getPermittedSettings`? Stephan added this for a reason
> [0].
>
> Also, when overriding a method from a parent class (as in case of
> `getDependencyDefinitions`) make sure to check those returning values
> before. For example, `parent::getDependencyDefinitions()` returns a [
> ... ] and you should merge them with added definitions.
>
> Why did you add `protected function runTestCaseFile(
> JsonTestCaseFileHandler $jsonTestCaseFileHandler ) { }`?
>
> `JsonTestCaseScriptRunnerTest` is inherited from [1], so it will call
> methods as needed except when they overridden. Your change actually
> makes all integration tests fail because `runTestCaseFile` isn't
> execute on any test because of the empty `{}`.
>
> Did you run those tests locally? You should have noticed some issues
> with those changes, I strongly suggest to run tests before posting a
> PR.
>
> [0]
> https://github.com/SemanticMediaWiki/SemanticResultFormats/commit/eea7866c917a766af8836c1b387ea7ced080985c
> [1]
> https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/JSONScript/JsonTestCaseScriptRunnerTest.php
>
> Cheers
>
> On 4/1/19, James HK <jamesin.hongkong.1@gmail.com> wrote:
>>> to have the ability to test SRF with Mermaid Extension in MW < 1.29 we
>>> need
>>> to set the manifest_version fo Extension:Mermaid to "1".
>>
>> SRF 3.1 will switch to MW 1.31 as minimum [0], so that shouldn't pose
>> any issue. The travis-CI matrix should be amended outside of the PR to
>> meet the 3.1 requirement.
>>
>> [0]
>> https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/docs/COMPATIBILITY.md#platform-compatibility-and-release-status
>>
>>> cc40a937de9156cb9316be574a26bc17ffee6236  add
>>> build/travis/install-mermaid.sh, extend JsonTestCaseScriptRunnerTest.php
>>
>> A quick look reveals:
>>
>> Not sure why you copied the `runTestCaseFile` method but that doesn't
>> belong here.
>>
>> ```
>> 	/**
>> 	 * @see JsonTestCaseScriptRunner::runTestCaseFile
>> 	 *
>> 	 * @param JsonTestCaseFileHandler $jsonTestCaseFileHandler
>> 	 */
>> 	protected function runTestCaseFile( JsonTestCaseFileHandler
>> $jsonTestCaseFileHandler ) {
>> 		$this->checkEnvironmentToSkipCurrentTest( $jsonTestCaseFileHandler );
>> 		// Setup
>> 		$this->prepareTest( $jsonTestCaseFileHandler );
>> 		// Before test execution
>> 		$this->doRunBeforeTest( $jsonTestCaseFileHandler );
>> 		// Run test cases
>> 		$this->doRunParserTests( $jsonTestCaseFileHandler );
>> 		$this->doRunSpecialTests( $jsonTestCaseFileHandler );
>> 		$this->doRunRdfTests( $jsonTestCaseFileHandler );
>> 		$this->doRunQueryTests( $jsonTestCaseFileHandler );
>> 		$this->doRunParserHtmlTests( $jsonTestCaseFileHandler );
>> 		$this->doRunApiTests( $jsonTestCaseFileHandler );
>> 	}
>> ```
>>
>> When you define a dependency, please make sure it does test against
>> something meaningful such as `MERMAID_VERSION` and not `SM_VERSION`.
>>
>> This should have failed locally for obvious reasons!
>>
>> ```
>> 	/**
>> 	 * @see JsonTestCaseScriptRunner::getDependencyDefinitions
>> 	 */
>> 	protected function getDependencyDefinitions() {
>> 		return [
>> 			'Mermaid' => function( $val, &$reason ) {
>> 				if ( !defined( 'SM_VERSION' ) ) {
>> 					$reason = "Dependency: Mermaid as requirement is not available!";
>> 					return false;
>> 				}
>> 				list( $compare, $requiredVersion ) = explode( ' ', $val );
>> 				$version = SM_VERSION;
>> 				if ( !version_compare( $version, $requiredVersion, $compare ) ) {
>> 					$reason = "Dependency: Required version of Mermaid
>> ($requiredVersion $compare $version) is not available!";
>> 					return false;
>> 				}
>> 				return true;
>> 			}
>> 		];
>> 	}
>> ```
>>
>> For better readability, you may write:
>>
>> ```
>> 	/**
>> 	 * @see JsonTestCaseScriptRunner::getDependencyDefinitions
>> 	 */
>> 	protected function getDependencyDefinitions() {
>> 		return [
>> 			'Mermaid' => [ $this, 'checkMermaidDependency' ]
>> 		];
>> 	}
>>
>> 	/**
>> 	 * Explain the reason ...
>> 	 */
>> 	public function checkMermaidDependency( $val, &$reason ) {
>>
>> 		if ( !defined( 'MERMAID_VERSION' ) ) {
>> 			$reason = "Dependency: Mermaid as requirement is not available!";
>> 			return false;
>> 		}
>>
>> 		list( $compare, $requiredVersion ) = explode( ' ', $val );
>> 		$version = MERMAID_VERSION;
>>
>> 		if ( !version_compare( $version, $requiredVersion, $compare ) ) {
>> 			$reason = "Dependency: Required version of Mermaid
>> ($requiredVersion $compare $version) is not available!";
>> 			return false;
>> 		}
>>
>> 		return true;
>> 	}
>> ```
>>
>> Cheers
>>
>> On 4/1/19, Sebastian Schmid <notifications@github.com> wrote:
>>> @mwjames: At the moment mermaid extension runs with manifest_version: 2.
>>> This just works with MW 1.29+. Tests execute with the following error
>>> msg:
>>>
>>> **Fatal error: Uncaught exception 'Exception' with message
>>> '/home/travis/build/RacoonDev/mw/extensions/Mermaid/extension.json:
>>> unsupported manifest_version: 2' in**
>>>
>>> to have the ability to test SRF with Mermaid Extension in MW < 1.29 we
>>> need
>>> to set the manifest_version fo Extension:Mermaid to "1".
>>>
>>> --
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly or view it on GitHub:
>>> https://github.com/SemanticMediaWiki/SemanticResultFormats/pull/467#issuecomment-478495422
>>
>

RacoonDev commented:
i turn crazy with this testing... but finally mermaid will be installed via composer in the build script. 
Now I'm facing a new problem... My tests are not passing the assertation [0]. The expected and actual strings are not identical. When i run the tests on my local system it passes all of my tests. I also tried to decode the actual sting but it just causes more and more error i cant get rid of. 
Could someone please have i look or give me a hint, I'm really stucking right now.

[0] https://travis-ci.org/SemanticMediaWiki/SemanticResultFormats/jobs/515644243

s7eph4n commented:
Will have a look tonight.

s7eph4n commented:
Turns out that CI tests were not run for any of the newer MW versions. "Newer" as in "any MW after 1.28".  :flushed: 

They were turned off at some point for a bug in MW [1] and never turned on again. In the meantime that bug has been fixed apparently, but the output of the MW parser changed, so when you started shuffling around test setups, tests started failing because they did not see their expected output anymore. So I fixed the failing tests (they where "mine" anyway) [2] and re-enabled testing on newer MW versions [3].

If you rebase your patch your tests should hopefully run. Or at least you should not get unrelated failures anymore.

[1] https://phabricator.wikimedia.org/T188840
[2] #483
[3] #484

RacoonDev commented:
@s7eph4n: Thank you for your look up and fixes! So i rebased my Fork and adjust the test environment. The line breaks in my gantt-01.json file were not correct so i changed it form CRLF to LF.

On my fork [1] every build job is successful and turns to green but on the PullRequest side [2] the tests are still failing. 

Can anybody tell my why? Am i overlooking something?

[1] https://travis-ci.org/RacoonDev/SemanticResultFormats
[2] https://travis-ci.org/SemanticMediaWiki/SemanticResultFormats/builds/518163605?utm_source=github_status&utm_medium=notification

mwjames commented:
The CI job tells you:

```
...PHP Deprecated:  Use of MagicWord::get was deprecated in MediaWiki
1.32. [Called from SMW\MediaWiki\MagicWordsFinder::findMagicWordInText
in
```

> Can anybody tell my why? Am i overlooking something?

Because some MW developer thought merging [0] without consequences to
third-party extensions goes unnoticed (without even trying to inform
extension developers that this sort of merge might cause failing CI
jobs) which is why the master job is failing with above message.

[0] https://github.com/wikimedia/mediawiki/commit/07628545608ec742dd21fd83f47b1552b898d3b4

On 4/10/19, Sebastian Schmid <notifications@github.com> wrote:
> @s7eph4n: Thank you for your look up and fixes! So i rebased my Fork and
> adjust the test environment. The line breaks in my gantt-01.json file were
> not correct so i changed it form CRLF to LF.
>
> On my fork [1] every build job is successful and turns to green but on the
> PullRequest side [2] the tests are still failing.
>
> Can anybody tell my why? Am i overlooking something?
>
> [1] https://travis-ci.org/RacoonDev/SemanticResultFormats
> [2]
> https://travis-ci.org/SemanticMediaWiki/SemanticResultFormats/builds/518163605?utm_source=github_status&utm_medium=notification
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/SemanticResultFormats/pull/467#issuecomment-481630873

s7eph4n commented:
Well, it's annoying, but to be fair, MW core devs have to deprecate stuff at some point and IMO it would not be reasonable to expect some kind of mail notification to every extension developer out there each time they deprecate something. SMW decided to leave Gerrit, so even if MW core were to run regular CI tests on extensions on Gerrit, there is no chance to catch failing SMW tests. As long as the deprecation policy is followed we should deal with it.
An idea might be to set `$wgDevelopmentWarnings = false` for tests triggered by PRs and pushes and `$wgDevelopmentWarnings = true` for cron jobs [1]. 

[1] https://docs.travis-ci.com/user/cron-jobs/#detecting-builds-triggered-by-cron

mwjames commented:
> Well, it's annoying, but to be fair, MW core devs have to deprecate stuff at some point and IMO it would not be reasonable to expect some kind of mail notification to every extension developer out there each time they deprecate something.

Of course, I'm personally annoyed, especially when you hit a triple like this weekend spending at least 2h on git blame and history to figure out what changed when and how.

> An idea might be to set $wgDevelopmentWarnings = false for tests triggered by PRs and pushes and $wgDevelopmentWarnings = true for cron jobs [1].

#485 made the MW master non voting.

mwjames commented:
> #485 made the MW master non voting.

@RacoonDev let's try this again. MW master is non voting now and after you have rebased your PR you should be able to make the tests to pass.

RacoonDev commented:
Okey, all tests have been passed. Is there yet anything that needs to be done before we can start merging?

@s7eph4n, @kghbln It tells me that there are still 2 changes requested. 
But I don't know how i can resolve those.

kghbln commented:
I will fix the I18n in a follow up.

kghbln commented on i18n/en.json:
```diff
@@ -337,5 +337,19 @@
+	"srf-paramdesc-gantt-barheight": "Height of task bars",
+	"srf-paramdesc-gantt-leftpadding": "Width of section title",
+	"srf-paramdesc-gantt-bargap": "Vertical distance of tasks bars from the margin",
+	"srf-error-gantt-mapping-assignment": "Wrong assignment of mapped values in '''$1'''",
```
@RacoonDev What is filled into the $1?

RacoonDev replied:
It's the mapping type like "statusmapping" or "prioritymapping"

kghbln replied:
Ok, thanks!

s7eph4n commented:
> 
> 
> Okey, all tests have been passed. Is there yet anything that needs to be done before we can start merging?
> 
> @s7eph4n, @kghbln It tells me that there are still 2 changes requested.
> But I don't know how i can resolve those.

No idea where those are coming from. 
As far as I am concerned, we can merge. Time permitting I may have a look again at the format and maybe suggest some changes.

RacoonDev commented:
@mwjames: just want to ask if we can expect a merge of this branch in the next few weeks? 
All checks have passed and all changes were approved... 
Or are there some open tasks we have to wait for?

kghbln commented:
This was actually not on James' list. :) The aim was to get this ready for SRF 3.1 so ...

RacoonDev commented:
@kghbln Thx
